lib/pure/asyncmacro.nim

```diff
@@ -214,9 +214,18 @@ proc asyncSingleProc(prc: NimNode): NimNode =
   # don't do anything with forward bodies (empty)
   if procBody.kind != nnkEmpty:
     # fix #13899, defer should not escape its original scope
-    procBody = newStmtList(newTree(nnkBlockStmt, newEmptyNode(), procBody))
-    procBody.add(createFutureVarCompletions(futureVarIdents, nil))
+    let blockStmt = newStmtList(newTree(nnkBlockStmt, newEmptyNode(), procBody))
+    procBody = newStmtList()
     let resultIdent = ident"result"
+    procBody.add quote do:
+      template setResult(x: `subRetType`) {.used.} =
```

Comment:
We've considered a similar thing in chronos so that nim-result can work better - https://github.com/status-im/nim-chronos/pull/297 - perhaps we should coordinate the mechanism here - one thing that often comes up is symbol resolution issues - ie template lookup scope is confusing / broken and has a tendency to ruin unrelated code when used together with generics: https://github.com/status-im/nim-stew/issues/161 - as a consequence, we've started using more "unique" names, also for "private" templates like this.

Reply:
From reading those PR's I think this is a different problem to whats in them. This doesn't effect how its returned but just means that code like this can now compile and work (Doesn't effect how normal `return` and `result = ` works)
```nim
# Not exactly a good async example, but just needed something basic
proc foo(x: int): Future[string] {.async.} =
  if x mod 2 == 0:
    "even"
  else:
    "odd"
```

But see what you mean about needing unique names. Maybe could make it very explicit like `asyncDispatchSetResult`?